pkg/operator/configobservation/apiserver/observe_apiserver.go

```diff
@@ -0,0 +1,276 @@
+package apiserver
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/imdario/mergo"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/configobservation"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+)
+
+const (
+	userServingCertPublicCertFile          = "/etc/kubernetes/static-pod-resources/secrets/user-serving-cert/tls.crt"
+	userServingCertPrivateKeyFile          = "/etc/kubernetes/static-pod-resources/secrets/user-serving-cert/tls.key"
+	namedUserServingCertResourceNameFormat = "user-serving-cert-%03d"
+)
+
+var namedUserServingCertResourceNames = []string{
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 0),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 1),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 2),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 3),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 4),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 5),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 6),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 7),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 8),
+	fmt.Sprintf(namedUserServingCertResourceNameFormat, 9),
+}
+
+var maxUserNamedCerts = len(namedUserServingCertResourceNames)
+
+// syncActionRules rules define source resource names indexed by destination resource names.
+// Empty value means to delete the destination.
+type syncActionRules map[string]string
+
+// resourceSyncFunc syncs a resource from the source location to the destination location.
+type resourceSyncFunc func(destination, source resourcesynccontroller.ResourceLocation) error
+
+// observeAPIServerConfigFunc observes configuration and returns the observedConfig and a map describing a list of
+// resources to sync.
+// It returns the observed config, sync rules and possibly an error. Nil sync rules mean to ignore all resources
+// in case of error. Otherwise, resources are deleted by default and the returned sync rules are taken as overrides of that.
+type observeAPIServerConfigFunc func(apiServer *configv1.APIServer, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (map[string]interface{}, syncActionRules, []error)
+
+// ObserveUserClientCABundle returns an ObserveConfigFunc that observes a user managed certificate bundle containing
+// signers that will be recognized for incoming client certificates in addition to the operator managed signers.
+var ObserveUserClientCABundle configobserver.ObserveConfigFunc = (&apiServerObserver{
```

Comment:
nit: no need for the var type

Reply:
i want to check the type safety :-)

Reply:
It will complain further down if types are wrong.